File: mixed/f1/group2/LongestPalindromicSubstring.java
// Given a string, return the longest palindromic substring.
// Sample 1:
// Input: s = "xyxyz"
// Output: "xyx", also "yxy"
// Time complexity: O(n^2)
// Space complexity: O(n)


class LongestPalindromicSubstring {
    public static void main(String...args) {
        String s = "xyxyz";
        System.out.println("Longest palindromic submistring: " + findLongestPalindromicSubstring(s));
    }

    static String findLongestPalindromicSubstring(String s) {
        int start = 0, end = 0;
        for (int i = 0; i < s.length(); i++) {
            int len1 = expandFromCenter(s, i, i+1);
            int len2 = expandFromCenter(s, i, i);
            int len = Math.max(len1, len2);
            if (end - start < len) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }

        return s.substring(start, end + 1);
    }

    static int expandFromCenter(String s, int i, int j) {
        while (i >= 0 && j < s.length() && s.charAt(i) == s.charAt(j)) {
            i--;
            j++;
        }
        return j - i - 1;
    }
}

File: bytopic/dpmultidimensional/python/group2/longest_palindromic_substring.py
"""
Given a string, return the longest palindromic substring.

Sample 1:
Input: s = "xyxyz"
Output: "xyx", also "yxy"

Tag: 145/150
Tag: 5/2927, R5/2936 (overall frequency ranking)
"""


def longest_palindrome_dp(s: str) -> str:
    if len(s) <= 1:
        return s
    max_len = 1
    max_str = s[0]
    dp = [[False for _ in range(len(s))] for _ in range(len(s))]
    for i in range(len(s)):
        dp[i][i] = True
        for j in range(i):
            if s[j] == s[i] and (i-j <= 2 or dp[j+1][i-1]):
                dp[j][i] = True
                if i-j+1 > max_len:
                    max_len = i-j+1
                    max_str = s[j:i+1]
    return max_str


def longest_palindrome_recursive_time_limit_exceeded(s: str) -> str:
    if s == s[::-1]:
        return s
    left = longest_palindrome_recursive_time_limit_exceeded(s[1:])
    right = longest_palindrome_recursive_time_limit_exceeded(s[:-1])

    if len(left) > len(right):
        return left
    else:
        return right


def main():
    s = "xyxyz"
    print(longest_palindrome_recursive_time_limit_exceeded(s))


if __name__ == "__main__":
    main()

